lib/synapse/haproxy.rb

```diff
@@ -1060,20 +1064,23 @@ def update_backends_at(socket_file_path, watchers)
 
       # parse the stats output to get current backends
       cur_backends = {}
+      re = Regexp.new('^(.+?),(.+?),(?:.*?,){15}(.+?),')
+
       info.split("\n").each do |line|
         next if line[0] == '#'
 
-        parts = line.split(',')
-        next if ['FRONTEND', 'BACKEND'].include?(parts[1])
+        name, addr, state = re.match(line)[1..3]
 
-        cur_backends[parts[0]] ||= []
-        cur_backends[parts[0]] << parts[1]
+        next if ['FRONTEND', 'BACKEND'].include?(address)
+
+        cur_backends[name] ||= {}
+        cur_backends[name][address] = state
       end
 
       # build a list of backends that should be enabled
       enabled_backends = {}
       watchers.each do |watcher|
-        enabled_backends[watcher.name] = []
+        enabled_backends[watcher.name] = {}
```

Comment:
if you're turning this into a hash to speed up lookups, i would recommend using a Set instead